mars/services/storage/spill.py

```diff
@@ -72,36 +74,91 @@ def unpin_data(self, key):
         if self._pinned_keys[key] <= 0:
             del self._pinned_keys[key]
 
+    @property
+    def spillable_size(self):
+        total_size = 0
+        for data_key, data_size in self._data_sizes.items():
+            if data_key not in self._pinned_keys and \
+                    data_key not in self._spilling_keys:
+                total_size += data_size
+        return total_size
+
     def get_spill_keys(self, size: int) -> Tuple[List, List]:
         spill_sizes = []
         spill_keys = []
         spill_size = 0
         for data_key, data_size in self._data_sizes.items():
+            if spill_size >= size:
+                break
             if data_key in self._pinned_keys:
                 continue
             if data_key in self._spilling_keys:
                 continue
             spill_sizes.append(data_size)
             spill_keys.append(data_key)
             spill_size += data_size
-            if spill_size > size:
-                break
+
         if spill_size < size:  # pragma: no cover
             pinned_sizes = dict((k, self._data_sizes[k]) for k in self._pinned_keys)
             spilling_keys = dict((k, self._data_sizes[k]) for k in self._spilling_keys)
-            raise NoDataToSpill(f'No data can be spilled for level: {self._level},'
-                                f'pinned keys: {pinned_sizes}, spilling keys: {spilling_keys}')
+            logger.debug('No data can be spilled for level: %s, pinned keys: %s,'
+                         ' spilling keys: %s', self._level, pinned_sizes, spilling_keys)
+            raise NoDataToSpill(f'No data can be spilled for level: {self._level}')
         self._spilling_keys.update(set(spill_keys))
         return spill_sizes, spill_keys
 
 
+class SpillManagerActor(mo.Actor):
+    """
+    The actor to handle the race condition when NoDataToSpill happens.
+    There are two situations when spill raises `NoDataToSpill`,
+    one is that space is allocated while objects are not put into storage,
+    another is some objects are pinned that can not be spilled,
+    so we create an asyncio event if not have enough objects to spill,
+    when put or unpin happens, we will notify and check spillable size,
+    if size is enough for spilling, call event.set() to wake up spilling task.
+    """
+    def __init__(self, level: StorageLevel):
+        self._level = level
+        self._event = None
+        self._lock = asyncio.Lock()
+
+    @classmethod
+    def gen_uid(cls, level: StorageLevel):
+        return f'spill_manager_{level}'
+
+    async def create_spill_event(self, size):
+        # make sure only one spilling task is waiting the event
+        yield self._lock.acquire()
+        self._event = event = asyncio.Event()
+        event.size = size
+
+    def notify_spill_event(self, spillable_size, quota_left):
```

Comment:
Rename to `notify_space_can_spill` or something like that. BTW, type annotation should be added, so do the other funcitons.